[NCCL] DDP communication hook: getFuture()

sinannasir opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #41840 [NCCL] [For Test] In DDP's reducer merge `work` and `future_work`
* **#41596 [NCCL] DDP communication hook: getFuture()**

We've modified the previous design of `convert_dist_work_to_future` API in the GH Issue [#39272](https://github.com/pytorch/pytorch/issues/39272).

1. Whenever we create a `WorkNCCL` object, create a `Future` associated with `WorkNCCL` and store it with the object.
2. Add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`.
3. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.
4. To mark the future associated with WorkNCCL completed, implement a `cudaStreamCallback` function.

`cudaStreamAddCallback` is marked as deprecated. An alternative is `cudaLaunchHostFunc`, but it is supported for CUDA > 10 and may not be deprecated until there's a reasonable alternative available according to [this discussion](https://stackoverflow.com/questions/56448390/how-to-recover-from-cuda-errors-when-using-cudalaunchhostfunc-instead-of-cudastr).

Differential Revision: [D22583690](https://our.internmc.facebook.com/intern/diff/D22583690/)

#### Test plan:
Run old  python test/distributed/test_c10d.py.
Some additional tests:
`test_ddp_comm_hook_allreduce_hook_nccl`: This unit test verifies whether a DDP communication hook that just calls allreduce gives the same result result with the case of no hook registered.  Without the then callback, the future_value in reducer is no longer a PyObject, and this unit test verifies future_value is properly checked.
`test_ddp_comm_hook_allreduce_then_mult_ten_hook_nccl`: This unit test verifies whether a DDP communication hook that calls allreduce and then multiplies the result by ten gives the expected result.

As of v10:
```
........................s.....s.....................................................s...............................
----------------------------------------------------------------------
Ran 116 tests

OK (skipped=3)
```

sinannasir committed:
[NCCL] DDP communication hook: getFuture()

We've modified the previous design of `convert_dist_work_to_future` API in the GH Issue [#39272](https://github.com/pytorch/pytorch/issues/39272).

1. Whenever we create a `WorkNCCL` object, create a `Future` associated with `WorkNCCL` and store it with the object.
2. Add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`.
3. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.
4. To mark the future associated with WorkNCCL completed, implement a `cudaStreamCallback` function.

`cudaStreamAddCallback` is marked as deprecated. An alternative is `cudaLaunchHostFunc`, but it is supported for CUDA > 10 and may not be deprecated until there's a reasonable alternative available according to [this discussion](https://stackoverflow.com/questions/56448390/how-to-recover-from-cuda-errors-when-using-cudalaunchhostfunc-instead-of-cudastr).

Differential Revision: [D22583690](https://our.internmc.facebook.com/intern/diff/D22583690/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture()"


We've modified the previous design of `convert_dist_work_to_future` API in the GH Issue [#39272](https://github.com/pytorch/pytorch/issues/39272).

1. Whenever we create a `WorkNCCL` object, create a `Future` associated with `WorkNCCL` and store it with the object.
2. Add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`.
3. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.
4. To mark the future associated with WorkNCCL completed, implement a `cudaStreamCallback` function.

`cudaStreamAddCallback` is marked as deprecated. An alternative is `cudaLaunchHostFunc`, but it is supported for CUDA > 10 and may not be deprecated until there's a reasonable alternative available according to [this discussion](https://stackoverflow.com/questions/56448390/how-to-recover-from-cuda-errors-when-using-cudalaunchhostfunc-instead-of-cudastr).

Differential Revision: [D22583690](https://our.internmc.facebook.com/intern/diff/D22583690/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture()"


We've modified the previous design of `convert_dist_work_to_future` API in the GH Issue [#39272](https://github.com/pytorch/pytorch/issues/39272).

1. Whenever we create a `WorkNCCL` object, create a `Future` associated with `WorkNCCL` and store it with the object.
2. Add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`.
3. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.
4. To mark the future associated with WorkNCCL completed, implement a `cudaStreamCallback` function.

`cudaStreamAddCallback` is marked as deprecated. An alternative is `cudaLaunchHostFunc`, but it is supported for CUDA > 10 and may not be deprecated until there's a reasonable alternative available according to [this discussion](https://stackoverflow.com/questions/56448390/how-to-recover-from-cuda-errors-when-using-cudalaunchhostfunc-instead-of-cudastr).

Differential Revision: [D22583690](https://our.internmc.facebook.com/intern/diff/D22583690/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture()"


We've modified the previous design of `convert_dist_work_to_future` API in the GH Issue [#39272](https://github.com/pytorch/pytorch/issues/39272).

1. Whenever we create a `WorkNCCL` object, create a `Future` associated with `WorkNCCL` and store it with the object.
2. Add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`.
3. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.
4. To mark the future associated with WorkNCCL completed, implement a `cudaStreamCallback` function.

`cudaStreamAddCallback` is marked as deprecated. An alternative is `cudaLaunchHostFunc`, but it is supported for CUDA > 10 and may not be deprecated until there's a reasonable alternative available according to [this discussion](https://stackoverflow.com/questions/56448390/how-to-recover-from-cuda-errors-when-using-cudalaunchhostfunc-instead-of-cudastr).

Differential Revision: [D22583690](https://our.internmc.facebook.com/intern/diff/D22583690/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture()"


We've modified the previous design of `convert_dist_work_to_future` API in the GH Issue [#39272](https://github.com/pytorch/pytorch/issues/39272).

1. Whenever we create a `WorkNCCL` object, create a `Future` associated with `WorkNCCL` and store it with the object.
2. Add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`.
3. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.
4. To mark the future associated with WorkNCCL completed, implement a `cudaStreamCallback` function.

`cudaStreamAddCallback` is marked as deprecated. An alternative is `cudaLaunchHostFunc`, but it is supported for CUDA > 10 and may not be deprecated until there's a reasonable alternative available according to [this discussion](https://stackoverflow.com/questions/56448390/how-to-recover-from-cuda-errors-when-using-cudalaunchhostfunc-instead-of-cudastr).

Differential Revision: [D22583690](https://our.internmc.facebook.com/intern/diff/D22583690/)

[ghstack-poisoned]

pritamdamania87 commented on test/distributed/test_c10d.py:
```diff
@@ -3075,6 +3075,35 @@ def test_ddp_comm_hook_future_passing_gpu_nccl(self):
+        )
+
+        def allreduce_hook(state: object, bucket: dist.GradBucket) -> torch.futures.Future:
+            return process_group.allreduce(bucket.get_tensors()).get_future()
```
nit: Maybe divide the result by some constant after the allreduce is done to ensure its different from the default result. This just ensures the hook is being called correctly and not the default implementation inadvertently.

sinannasir replied:
I've included a then callback within allreduce_hook.

pritamdamania87 commented on torch/lib/c10d/ProcessGroup.cpp:
```diff
@@ -55,8 +55,11 @@ bool ProcessGroup::Work::wait(std::chrono::milliseconds timeout) {
 
-void ProcessGroup::Work::abort() {
-  TORCH_CHECK(false, "ProcessGroup::Work::abort not implemented.")
+void ProcessGroup::Work::abort(){
+    TORCH_CHECK(false, "ProcessGroup::Work::abort not implemented.")}
```
Looks like the formatting messed up here.

sinannasir replied:
VS Code fixed that automatically. If I do 
```
void ProcessGroup::Work::abort(){
    TORCH_CHECK(false, "ProcessGroup::Work::abort not implemented.")
}
```
It gives a clang format warning.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -6,11 +6,14 @@
 
 #include <c10d/Utils.hpp>
 
+#include <torch/csrc/jit/python/pybind_utils.h>
```
Do we need to include this? We are adding a python dependency here.

sinannasir replied:
I guess so, because we need `PythonFutureWrapper` to bind Future object.

sinannasir replied:
No we don't, I forgot to remove that sorry :)

pritamdamania87 commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -89,6 +89,8 @@ class ProcessGroup {
 
     virtual void abort();
 
+    virtual c10::intrusive_ptr<c10::ivalue::Future> getFuture();
```
Add some docs for this method.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -704,6 +759,13 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
+      cudaStreamAddCallback(
+          ncclStream,
+          cudaStreamCallback,
+          (void*)std::make_shared<checkFuture>(
+              work, outputs, work_counter, inputs.size(), mutex_)
+              .get(),
```
You probably need to use the same `checkFuture` object for all the inputs so that `work_counter` is shared among them.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -660,6 +671,48 @@ std::shared_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
+  checkFuture(
+      std::shared_ptr<ProcessGroupNCCL::WorkNCCL> work,
+      std::vector<at::Tensor>& outputs,
+      int work_counter,
+      int input_size,
+      std::mutex& mutex)
```
Avoid passing `work_counter` and `mutex` here. `work_counter` can be initialized to 0 in the constructor and can be a `std::atomic`.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -660,6 +671,48 @@ std::shared_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
+        mutex_(mutex){};
+
+ public:
+  std::shared_ptr<ProcessGroupNCCL::WorkNCCL> work_;
+  std::vector<at::Tensor>& outputs_;
```
This should be private.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -660,6 +671,48 @@ std::shared_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
   return std::make_shared<ProcessGroupNCCL::WorkNCCL>(devices);
 }
 
+struct checkFuture {
```
nit: Start name with capital.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -660,6 +671,48 @@ std::shared_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
+      int work_counter,
+      int input_size,
+      std::mutex& mutex)
+      : work_(std::move(work)),
```
Just pass the future to this class since you need the future and nothing else from `work_`.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -660,6 +671,48 @@ std::shared_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
+ public:
+  std::shared_ptr<ProcessGroupNCCL::WorkNCCL> work_;
+  std::vector<at::Tensor>& outputs_;
+  bool markFutureCompletedCheck() {
+    std::unique_lock<std::mutex> lock(mutex_);
+    work_counter_ += 1;
+    return work_counter_ == input_size_;
+  };
+  void markFutureCompleted() {
+    at::launch(([this] {
+      work_->getFuture()->markCompleted(torch::jit::IValue(outputs_));
+    }));
+  }
```
Might be simpler to combine this into a single function:

```
void markWorkCompleted() {
    at::launch(([this] {
      work_counter_++;
      if (work_counter_ == input_size) {
          future_->markCompleted(torch::jit::IValue(outputs_));
      }
    }));
}
```

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -660,6 +671,48 @@ std::shared_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
+    cudaStream_t /* unused */,
+    cudaError_t /* unused */,
+    void* userData) {
+  auto hostfuncdata = (checkFuture*)userData;
```
Use `std::dynamic_pointer_cast` and verify the return value of dynamic_pointer_cast is not null.

sinannasir replied:
@pritamdamania87 do you mean smth like `dynamic_cast<CheckFutureWork*>((CheckFutureWork*)userData)`. 

This is an issue now. I think I am not properly casting, because even a simple `workCounter_` increment and `std::cout`, prints meaningless stuff. Looks like object is corrupted. Probably because of the way I'm casting it?

pritamdamania87 replied:
I meant `auto hostfuncdata = std::dynamic_cast<CheckFutureWork*>(userData)`.

sinannasir replied:
@pritamdamania87 I see, the way that I'm doing that cast doesn't provide any protection. Thanks!

Following the discussion in https://stackoverflow.com/questions/55006471/safe-way-to-dynamic-cast-void-to-a-type , I'll just create a base class of CheckFutureWork and `static_cast` to it first and finally do `dynamic_cast` like: 
```
  auto castedUserDataBase = static_cast<CheckFutureWorkBase*>(userData);
  auto castedUserData = dynamic_cast<CheckFutureWork*>(castedUserDataBase);
```
Let me know if a base class is not necessarily needed for our case.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -660,6 +671,48 @@ std::shared_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
+  std::mutex& mutex_;
+};
+
+void cudaStreamCallback(
```
nit: Let's call this `ncclKernelCompletionCallback`.

pritamdamania87 reviewed (commented):
I understand the PR is not yet completely ready, but just leaving some early comments that would be helpful when you post a review ready version.

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture()"


We've modified the previous design of `convert_dist_work_to_future` API in the GH Issue [#39272](https://github.com/pytorch/pytorch/issues/39272).

1. Whenever we create a `WorkNCCL` object, create a `Future` associated with `WorkNCCL` and store it with the object.
2. Add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`.
3. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.
4. To mark the future associated with WorkNCCL completed, implement a `cudaStreamCallback` function.

`cudaStreamAddCallback` is marked as deprecated. An alternative is `cudaLaunchHostFunc`, but it is supported for CUDA > 10 and may not be deprecated until there's a reasonable alternative available according to [this discussion](https://stackoverflow.com/questions/56448390/how-to-recover-from-cuda-errors-when-using-cudalaunchhostfunc-instead-of-cudastr).

Differential Revision: [D22583690](https://our.internmc.facebook.com/intern/diff/D22583690/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture()"


We've modified the previous design of `convert_dist_work_to_future` API in the GH Issue [#39272](https://github.com/pytorch/pytorch/issues/39272).

1. Whenever we create a `WorkNCCL` object, create a `Future` associated with `WorkNCCL` and store it with the object.
2. Add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`.
3. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.
4. To mark the future associated with WorkNCCL completed, implement a `cudaStreamCallback` function.

`cudaStreamAddCallback` is marked as deprecated. An alternative is `cudaLaunchHostFunc`, but it is supported for CUDA > 10 and may not be deprecated until there's a reasonable alternative available according to [this discussion](https://stackoverflow.com/questions/56448390/how-to-recover-from-cuda-errors-when-using-cudalaunchhostfunc-instead-of-cudastr).

Differential Revision: [D22583690](https://our.internmc.facebook.com/intern/diff/D22583690/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture()"


We've modified the previous design of `convert_dist_work_to_future` API in the GH Issue [#39272](https://github.com/pytorch/pytorch/issues/39272).

1. Whenever we create a `WorkNCCL` object, create a `Future` associated with `WorkNCCL` and store it with the object.
2. Add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`.
3. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.
4. To mark the future associated with WorkNCCL completed, implement a `cudaStreamCallback` function.

`cudaStreamAddCallback` is marked as deprecated. An alternative is `cudaLaunchHostFunc`, but it is supported for CUDA > 10 and may not be deprecated until there's a reasonable alternative available according to [this discussion](https://stackoverflow.com/questions/56448390/how-to-recover-from-cuda-errors-when-using-cudalaunchhostfunc-instead-of-cudastr).

Differential Revision: [D22583690](https://our.internmc.facebook.com/intern/diff/D22583690/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture()"


We've modified the previous design of `convert_dist_work_to_future` API in the GH Issue [#39272](https://github.com/pytorch/pytorch/issues/39272).

1. Whenever we create a `WorkNCCL` object, create a `Future` associated with `WorkNCCL` and store it with the object.
2. Add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`.
3. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.
4. To mark the future associated with WorkNCCL completed, implement a `cudaStreamCallback` function.

`cudaStreamAddCallback` is marked as deprecated. An alternative is `cudaLaunchHostFunc`, but it is supported for CUDA > 10 and may not be deprecated until there's a reasonable alternative available according to [this discussion](https://stackoverflow.com/questions/56448390/how-to-recover-from-cuda-errors-when-using-cudalaunchhostfunc-instead-of-cudastr).

Differential Revision: [D22583690](https://our.internmc.facebook.com/intern/diff/D22583690/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture()"


We've modified the previous design of `convert_dist_work_to_future` API in the GH Issue [#39272](https://github.com/pytorch/pytorch/issues/39272).

1. Whenever we create a `WorkNCCL` object, create a `Future` associated with `WorkNCCL` and store it with the object.
2. Add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`.
3. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.
4. To mark the future associated with WorkNCCL completed, implement a `cudaStreamCallback` function.

`cudaStreamAddCallback` is marked as deprecated. An alternative is `cudaLaunchHostFunc`, but it is supported for CUDA > 10 and may not be deprecated until there's a reasonable alternative available according to [this discussion](https://stackoverflow.com/questions/56448390/how-to-recover-from-cuda-errors-when-using-cudalaunchhostfunc-instead-of-cudastr).

Differential Revision: [D22583690](https://our.internmc.facebook.com/intern/diff/D22583690/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -697,13 +721,27 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
+      // call ncclKernelCompletionCallback once the appropriate NCCL kernel
+      // is done and workNCCL's Future work will be marked as completed
+      // using this callback.
+      cudaStreamAddCallback(
```
I think the outputs not being populated issue might come from this: https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/groups.html
"Caution: When called inside a group, stream operations (like ncclAllReduce) can return without having enqueued the operation on the stream. Stream operations like cudaStreamSynchronize can therefore be called only after ncclGroupEnd returns."

Maybe we need to move this cudaStreamAddCallback calls outside of the `nccl_group_guard`?

sinannasir replied:
I'm trying that. Thanks,

rohan-varma replied:
The [docs] for `cudaStreamAddCallback` say "Adds a callback to be called on the host after all currently enqueued items in the stream have completed", so my reasoning in conjunction with the above link was that if the operation is not enqueued, then there is a chance, according to the documentation, that the callback would fire, because the condition for the callbacks being called is (vacuously) true. Although, I may not be interpreting the docs correctly and maybe this can't happen in practice.

rohan-varma commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -8,11 +8,69 @@
+      if (streamCounter_ == (*outputs_).size()) {
+        // Need to wait until work is completed, because some post processing
+        // could be needed for outputs.
+        work_->wait();
```
I think we might be able to get rid of this if we consider the group-guard related change I mentioned below.

sinannasir replied:
replaced with `synchronizeStreams`

pritamdamania87 commented on test/distributed/test_c10d.py:
```diff
@@ -3048,32 +3052,86 @@ def fut_then(fut):
+
+    @requires_nccl()
+    @skip_if_lt_x_gpu(2)
+    def test_ddp_comm_hook_allreduce_then_mult_ten_hook_nccl(self):
```
nit: s/ten/then

sinannasir replied:
ten is actually 10, I'll just replace this as `test_ddp_comm_hook_allreduce_then_mult_hook_nccl`

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -700,6 +701,13 @@ They are used in specifying strategies for reduction collectives, e.g.,
           py::arg("timeout") = kNoTimeout,
+          py::call_guard<py::gil_scoped_release>())
+      .def(
+          "get_future",
```
Add some docs for this method since it is a public method. Example: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/init.cpp#L291

sinannasir replied:
Should we make it private like `_get_future()` as well?

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -643,6 +646,9 @@ def _register_comm_hook(self, state: object, hook: callable):
         .. warning ::
             DDP communication hook is experimental and subject to change.
 
+        .. warning ::
+            "get_future" API supports only NCCL backend.
+
```
This should be added as part of the docs for `get_future` and not here.

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture()"


We've modified the previous design of `convert_dist_work_to_future` API in the GH Issue [#39272](https://github.com/pytorch/pytorch/issues/39272).

1. Whenever we create a `WorkNCCL` object, create a `Future` associated with `WorkNCCL` and store it with the object.
2. Add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`.
3. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.
4. To mark the future associated with WorkNCCL completed, implement a `cudaStreamCallback` function.

`cudaStreamAddCallback` is marked as deprecated. An alternative is `cudaLaunchHostFunc`, but it is supported for CUDA > 10 and may not be deprecated until there's a reasonable alternative available according to [this discussion](https://stackoverflow.com/questions/56448390/how-to-recover-from-cuda-errors-when-using-cudalaunchhostfunc-instead-of-cudastr).

Differential Revision: [D22583690](https://our.internmc.facebook.com/intern/diff/D22583690/)

[ghstack-poisoned]

sinannasir commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -8,11 +8,69 @@
+            "completed by ncclKernelCompletionCallback.")
+        work_->getFuture()->markCompleted(at::IValue(*outputs_));
+
+        checkFutObjs_->erase(getPtr());
```
Yes, I was using `std::atomic` for `streamCounter_` and locking `checkFutObjMutex_` in this line, but that gave `future can't be marked completed twice error`, so I thought having one lock for both is safer. That error never appeared again, so I thought that was the reason :) 

Unfortunately, I didn't record the error. let me try to duplicate the error with `std::atomic` case.

sinannasir replied:
Okay, just reproduced the error with `std::atomic<int> checkCounter_` and locking `checkFutObjMutex_` at line 59:
```
........................s.....s.....................................................s......[E thread_pool.cpp:112] Exception in thread pool task: Future work of workNCCL can only be marked as completed by ncclKernelCompletionCallback.
Exception raised from operator() at ../torch/lib/c10d/../c10d/ProcessGroupNCCL.hpp:179 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x4d (0x7fedb075977d in /home/sinannasir/local/pytorch/torch/lib/libc10.so)
frame #1: <unknown function> + 0xb59dc6 (0x7fedcb40adc6 in /home/sinannasir/local/pytorch/torch/lib/libtorch_python.so)
frame #2: <unknown function> + 0x7eead8 (0x7fedc4002ad8 in /home/sinannasir/local/pytorch/torch/lib/libtorch_cpu.so)
frame #3: c10::ThreadPool::main_loop(unsigned long) + 0x283 (0x7fedb0755383 in /home/sinannasir/local/pytorch/torch/lib/libc10.so)
frame #4: <unknown function> + 0xc819d (0x7fedcba5c19d in /home/sinannasir/local/miniconda3/envs/pytorch/lib/libstdc++.so.6)
frame #5: <unknown function> + 0x7ea5 (0x7fede75a0ea5 in /lib64/libpthread.so.0)
frame #6: clone + 0x6d (0x7fede72c98dd in /lib64/libc.so.6)

.........................
----------------------------------------------------------------------
Ran 116 tests
```
Let me try to remove `TORCH_CHECK` to get more info.

sinannasir replied:
Thanks @pritamdamania87  for explaining this offline.

sinannasir commented on torch/csrc/distributed/c10d/comm.cpp:
```diff
@@ -113,14 +113,18 @@ c10::intrusive_ptr<torch::jit::Future> PythonCommHook::runHook(
-
-  return value.toTensorVector();
+  // Since we have a Python hook, future_value can be a PyObject.
+  if (future_value.isPyObject()) {
```
Yes, this is checked by 
https://github.com/pytorch/pytorch/blob/fe887ee17cad10b0c37823a0e2e309574b7c141d/test/distributed/test_c10d.py#L3091-L3108

When we don't have `then` callback without this condition we would get:
```
RuntimeError: isPyObject() INTERNAL ASSERT FAILED at "../aten/src/ATen/core/ivalue_inl.h":103, please report a bug to PyTorch. Expected PyObject but gotGenericList
Exception raised from toPyObjectHolder at ../aten/src/ATen/core/ivalue_inl.h:103 (most recent call first):
```

sinannasir commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -8,11 +8,69 @@
+ private:
+  std::shared_ptr<ProcessGroup::Work> work_;
+  std::shared_ptr<std::vector<at::Tensor>> outputs_;
+  int streamCounter_;
```
please see this thread https://github.com/pytorch/pytorch/pull/41596#discussion_r459107787

sinannasir commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -8,11 +8,69 @@
+  }
+
+  void markFutureCompleted() {
+    at::launch(([this]() {
```
Discussed this offline and decided that it is fine to pass `this` since we store a `shared_ptr` to this inside `ProcessGroup::checkFutObjs_`.

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture()"


We've modified the previous design of `convert_dist_work_to_future` API in the GH Issue [#39272](https://github.com/pytorch/pytorch/issues/39272).

1. Whenever we create a `WorkNCCL` object, create a `Future` associated with `WorkNCCL` and store it with the object.
2. Add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`.
3. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.
4. To mark the future associated with WorkNCCL completed, implement a `cudaStreamCallback` function.

`cudaStreamAddCallback` is marked as deprecated. An alternative is `cudaLaunchHostFunc`, but it is supported for CUDA > 10 and may not be deprecated until there's a reasonable alternative available according to [this discussion](https://stackoverflow.com/questions/56448390/how-to-recover-from-cuda-errors-when-using-cudalaunchhostfunc-instead-of-cudastr).

Differential Revision: [D22583690](https://our.internmc.facebook.com/intern/diff/D22583690/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture()"


We've modified the previous design of `convert_dist_work_to_future` API in the GH Issue [#39272](https://github.com/pytorch/pytorch/issues/39272).

1. Whenever we create a `WorkNCCL` object, create a `Future` associated with `WorkNCCL` and store it with the object.
2. Add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`.
3. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.
4. To mark the future associated with WorkNCCL completed, implement a `cudaStreamCallback` function.

`cudaStreamAddCallback` is marked as deprecated. An alternative is `cudaLaunchHostFunc`, but it is supported for CUDA > 10 and may not be deprecated until there's a reasonable alternative available according to [this discussion](https://stackoverflow.com/questions/56448390/how-to-recover-from-cuda-errors-when-using-cudalaunchhostfunc-instead-of-cudastr).

Differential Revision: [D22583690](https://our.internmc.facebook.com/intern/diff/D22583690/)

[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture()"


We've modified the previous design of `convert_dist_work_to_future` API in the GH Issue [#39272](https://github.com/pytorch/pytorch/issues/39272).

1. Whenever we create a `WorkNCCL` object, create a `Future` associated with `WorkNCCL` and store it with the object.
2. Add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`.
3. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.
4. To mark the future associated with WorkNCCL completed, implement a `cudaStreamCallback` function.

`cudaStreamAddCallback` is marked as deprecated. An alternative is `cudaLaunchHostFunc`, but it is supported for CUDA > 10 and may not be deprecated until there's a reasonable alternative available according to [this discussion](https://stackoverflow.com/questions/56448390/how-to-recover-from-cuda-errors-when-using-cudalaunchhostfunc-instead-of-cudastr).

Differential Revision: [D22583690](https://our.internmc.facebook.com/intern/diff/D22583690/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -708,7 +708,18 @@ They are used in specifying strategies for reduction collectives, e.g.,
+          R"(
+            ``get_future`` retrieves a future associated with the completion of
+            ``c10d.ProcessGroup.work``. As an example, a future object can be set
+            by `future_work = dist.allreduce(bucket.get_tensors()).get_tensors()`
```
The last part in this line should be `get_future` instead of `get_tensors`? Also, in the example just use something like `dist.allreduce(tensor)`, no need to mention GradBucket here since there is a generic API for work.

pritamdamania87 commented:
Looks good, although please address my last comment before landing.

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture()"


We've modified the previous design of `convert_dist_work_to_future` API in the GH Issue [#39272](https://github.com/pytorch/pytorch/issues/39272).

1. Whenever we create a `WorkNCCL` object, create a `Future` associated with `WorkNCCL` and store it with the object.
2. Add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`.
3. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.
4. To mark the future associated with WorkNCCL completed, implement a `cudaStreamCallback` function.

`cudaStreamAddCallback` is marked as deprecated. An alternative is `cudaLaunchHostFunc`, but it is supported for CUDA > 10 and may not be deprecated until there's a reasonable alternative available according to [this discussion](https://stackoverflow.com/questions/56448390/how-to-recover-from-cuda-errors-when-using-cudalaunchhostfunc-instead-of-cudastr).

Differential Revision: [D22583690](https://our.internmc.facebook.com/intern/diff/D22583690/)

#### Test plan:
Run old  python test/distributed/test_c10d.py.
Some additional tests:
`test_ddp_comm_hook_allreduce_hook_nccl`: This unit test verifies whether a DDP communication hook that just calls allreduce gives the same result result with the case of no hook registered.  Without the then callback, the future_value in reducer is no longer a PyObject, and this unit test verifies future_value is properly checked.
`test_ddp_comm_hook_allreduce_then_mult_ten_hook_nccl`: This unit test verifies whether a DDP communication hook that calls allreduce and then multiplies the result by ten gives the expected result.

As of v10:
```
........................s.....s.....................................................s...............................
----------------------------------------------------------------------
Ran 116 tests

OK (skipped=3)
```
`flow-cli` performance validation using a stacked diff when bucket.work is completely replaced with bucket.future_work in `reducer`. See PR [#41840](https://github.com/pytorch/pytorch/pull/41840) [D22660198](https://www.internalfb.com/intern/diff/D22660198/).


[ghstack-poisoned]

sinannasir committed:
Update on "[NCCL] DDP communication hook: getFuture()"


We've modified the previous design of `convert_dist_work_to_future` API in the GH Issue [#39272](https://github.com/pytorch/pytorch/issues/39272).

1. Whenever we create a `WorkNCCL` object, create a `Future` associated with `WorkNCCL` and store it with the object.
2. Add an API `c10::intrusive_ptr<c10::ivalue::Future> getFuture()` to `c10d::ProcessGroup::Work`.
3. This API will only be supported by NCCL in the first version, the default implementation will throw UnsupportedOperation.
4. To mark the future associated with WorkNCCL completed, implement a `cudaStreamCallback` function.

`cudaStreamAddCallback` is marked as deprecated. An alternative is `cudaLaunchHostFunc`, but it is supported for CUDA > 10 and may not be deprecated until there's a reasonable alternative available according to [this discussion](https://stackoverflow.com/questions/56448390/how-to-recover-from-cuda-errors-when-using-cudalaunchhostfunc-instead-of-cudastr).

Differential Revision: [D22583690](https://our.internmc.facebook.com/intern/diff/D22583690/)

#### Test plan:
Run old  python test/distributed/test_c10d.py.
Some additional tests:
`test_ddp_comm_hook_allreduce_hook_nccl`: This unit test verifies whether a DDP communication hook that just calls allreduce gives the same result result with the case of no hook registered.  Without the then callback, the future_value in reducer is no longer a PyObject, and this unit test verifies future_value is properly checked.
`test_ddp_comm_hook_allreduce_then_mult_ten_hook_nccl`: This unit test verifies whether a DDP communication hook that calls allreduce and then multiplies the result by ten gives the expected result.

As of v10:
```
........................s.....s.....................................................s...............................
----------------------------------------------------------------------
Ran 116 tests

OK (skipped=3)
```
`flow-cli` performance validation using a stacked diff where bucket.work is completely replaced with bucket.future_work in `reducer`. See PR [#41840](https://github.com/pytorch/pytorch/pull/41840) [D22660198](https://www.internalfb.com/intern/diff/D22660198/).


[ghstack-poisoned]
